messages2/src/protocols/cred_issuance/mod.rs

```diff
@@ -0,0 +1,170 @@
+            },
+            traits::MessageKind,
+        },
+        MessageType, Protocol,
+    },
+    protocols::traits::DelayedSerde,
+};
+
+#[derive(Clone, Debug, From, PartialEq)]
+pub enum CredentialIssuance {
+    OfferCredential(OfferCredential),
+    ProposeCredential(ProposeCredential),
+    RequestCredential(RequestCredential),
+    IssueCredential(IssueCredential),
+    Ack(AckCredential),
+}
+
+impl DelayedSerde for CredentialIssuance {
+    type MsgType<'a> = (CredentialIssuanceKind, &'a str);
+
+    fn delayed_deserialize<'de, D>(msg_type: Self::MsgType<'de>, deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let (major, kind) = msg_type;
+        let CredentialIssuanceKind::V1(major) = major;
+        let CredentialIssuanceV1::V1_0(_minor) = major;
+        let kind = CredentialIssuanceV1_0Kind::from_str(kind).map_err(D::Error::custom)?;
+
+        match kind {
+            CredentialIssuanceV1_0Kind::OfferCredential => {
+                OfferCredential::delayed_deserialize(kind, deserializer).map(From::from)
+            }
+            CredentialIssuanceV1_0Kind::ProposeCredential => {
+                ProposeCredential::delayed_deserialize(kind, deserializer).map(From::from)
+            }
+            CredentialIssuanceV1_0Kind::RequestCredential => {
+                RequestCredential::delayed_deserialize(kind, deserializer).map(From::from)
+            }
+            CredentialIssuanceV1_0Kind::IssueCredential => {
+                IssueCredential::delayed_deserialize(kind, deserializer).map(From::from)
+            }
+            CredentialIssuanceV1_0Kind::Ack => AckCredential::delayed_deserialize(kind, deserializer).map(From::from),
+            CredentialIssuanceV1_0Kind::CredentialPreview => Err(utils::not_standalone_msg::<D>(kind.as_ref())),
+        }
+    }
+
+    fn delayed_serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::OfferCredential(v) => v.delayed_serialize(serializer),
+            Self::ProposeCredential(v) => v.delayed_serialize(serializer),
+            Self::RequestCredential(v) => v.delayed_serialize(serializer),
+            Self::IssueCredential(v) => v.delayed_serialize(serializer),
+            Self::Ack(v) => v.delayed_serialize(serializer),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct CredentialPreview {
+    #[serde(rename = "@type")]
+    msg_type: CredentialPreviewMsgType,
+    pub attributes: Vec<CredentialAttr>,
+}
+
+impl CredentialPreview {
+    pub fn new(attributes: Vec<CredentialAttr>) -> Self {
+        Self {
+            msg_type: CredentialPreviewMsgType,
+            attributes,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+struct CredentialPreviewMsgType;
+
+impl From<CredentialPreviewMsgType> for CredentialIssuanceV1_0Kind {
```

Comment:
What is the purpose of this impl? Seems unused.

Reply:
It is. Looks like a leftover from before the `MessageType` refactoring into `Protocol` and the other types representing the message kind. Great catch!

In fact, this gives me an idea that could make serialization/deserialization easier for these types (that don't represent a standalone message type).

Reply:
Simplified deserialization for these *non-standalone* message types. The `From` impls stayed as a result.